pilot/pkg/model/service.go

```diff
@@ -488,18 +488,35 @@ func (h Hostname) String() string {
 //  Hostname("*.com").Matches("foo.com")     = true
 //  Hostname("*.foo.com").Matches("foo.com") = false
 func (h Hostname) Matches(o Hostname) bool {
-	if !strings.Contains(string(h), "*") && !strings.Contains(string(o), "*") {
+	if len(h) == 0 && len(o) == 0 {
+		return true
+	}
+
+	hWildcard := string(h[0]) == "*"
+	if hWildcard && len(o) == 0 {
+		return true
```

Comment:
how is this true? *.com is not equal to ""

Reply:
@ZackButcher tweaked this code so I defer to him. All the tests pass so are we missing a test here?

Reply:
I think you added in this clause when you swapped to retrieving the wildcard by index, to avoid checking `o[0]` when `len(o) == 0`. IMO the empty hostname should match nothing, so I think we can just update the code to:

<pre>func (h Hostname) Matches(o Hostname) bool {
    if <b>len(h) == 0 || len(o) == 0</b> {
        return <b>false</b>
    }
    // ...
}
</pre>

Then we know that both hostnames are at least one char long and can do our indexing, removing this new if that's incorrect.

And we should add a test case for `Hostname("*.com").Matches("")` (which I _thought_ we had already?).